netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +97,252 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
   }
 
   /**
-   * Creates an TLS negotiation for HTTP/2 using ALPN/NPN.
+   * Returns a {@link ChannelHandler} that ensures the pipeline is set up so that TLS will
+   * be negotiated, the {@code handler} is added and writes to the {@link Channel} may happen
+   * immediately, even before the TLS Handshake is complete.
    */
-  public static Negotiation tls(final SSLEngine sslEngine, final ChannelHandler... handlers) {
-    Preconditions.checkArgument(handlers.length > 0, "No handlers were provided");
+  public static ChannelHandler tls(final SSLEngine sslEngine, final ChannelHandler handler) {
     Preconditions.checkNotNull(sslEngine, "sslEngine");
-
     final SettableFuture<Void> completeFuture = SettableFuture.create();
     if (!installJettyTlsProtocolSelection(sslEngine, completeFuture, false)) {
       throw new IllegalStateException("NPN/ALPN extensions not installed");
     }
-    final ChannelInitializer<Channel> initializer = new ChannelInitializer<Channel>() {
-      @Override
-      public void initChannel(final Channel ch) throws Exception {
-        SslHandler sslHandler = new SslHandler(sslEngine, false);
-        sslHandler.handshakeFuture().addListener(
-            new GenericFutureListener<Future<? super Channel>>() {
-              @Override
-              public void operationComplete(Future<? super Channel> future) throws Exception {
-                // If an error occurred during the handshake, throw it
-                // to the pipeline.
-                java.util.concurrent.Future<?> doneFuture =
-                    future.isSuccess() ? completeFuture : future;
-                doneFuture.get();
-              }
-            });
-        ch.pipeline().addLast(sslHandler);
-        ch.pipeline().addLast(handlers);
-      }
-    };
-
-    return new Negotiation() {
-      @Override
-      public ChannelInitializer<Channel> initializer() {
-        return initializer;
-      }
-
-      @Override
-      public void onConnected(Channel channel) {
-        // Nothing to do.
-      }
-
-      @Override
-      public ListenableFuture<Void> completeFuture() {
-        return completeFuture;
-      }
-    };
+    SslHandler sslHandler = new SslHandler(sslEngine, false);
+    sslHandler.handshakeFuture().addListener(
+        new GenericFutureListener<Future<? super Channel>>() {
+          @Override
+          public void operationComplete(Future<? super Channel> future) throws Exception {
+            // If an error occurred during the handshake, throw it to the pipeline.
+            if (future.isSuccess()) {
+              completeFuture.get();
```

Comment:
@nmittler just talked with Eric about it. We "probably possibly" need it :-). Suggested to leave it untouched for now, all this ALPN related stuff will be deleted once with switch to Netty's ALPN handling and afaik this is due for the 0.5 release. So very soon.